app/src/main/java/com/example/caculatorapp/MainActivity.kt

```diff
@@ -2,10 +2,193 @@ package com.example.caculatorapp
+                    canAddOperation = false
+                    alreadyAddDemical = true
+                }
+            } else {
+                binding.tvWorkings.append(view.text)
+                canAddOperation = true
+                if (!alreadyAddDemical) canAddDemical = true
+            }
+        }
+    }
+
+    fun operatorAction(view: View) {
+        if (binding.tvWorkings.length() > 0 && view is Button && canAddOperation && isNumber(binding.tvWorkings.text.last())) {
+            binding.tvWorkings.append(view.text)
+            if (alreadyAddDemical) alreadyAddDemicalPrevious = true
+            else alreadyAddDemicalPrevious = false
+            canAddOperation = false
+            alreadyAddDemical = false
+            canAddDemical = false
+        }
+    }
+
+    fun allClearAction(view: View) {
+        binding.tvWorkings.text = ""
+        binding.tvResults.text = ""
+        canAddOperation = false
+        canAddDemical = false
+        alreadyAddDemical = false
+        alreadyAddDemicalPrevious = false
+    }
+
+    fun backspaceAction(view: View) {
+        val length = binding.tvWorkings.length()
+        if (length > 0) {
+            var lastToRemove: Char = binding.tvWorkings.text.last()
+            when (lastToRemove) {
+                '+', '-', 'x', '/' -> {
+                    binding.tvWorkings.text = binding.tvWorkings.text.subSequence(0, length - 1)
+                    canAddOperation = true
+                    if (!alreadyAddDemicalPrevious) canAddDemical = true
+                }
+                '.' -> {
+                    binding.tvWorkings.text = binding.tvWorkings.text.subSequence(0, length - 1)
+                    canAddOperation = true
+                    canAddDemical = true
+                    alreadyAddDemical = false
+                }
+                else -> binding.tvWorkings.text = binding.tvWorkings.text.subSequence(0, length - 1)
+            }
+        }
+
+    }
+
+    fun equalsAction(view: View) {
+        binding.tvResults.text = caculateResults()
+    }
+
+    private fun isNumber(char: Char): Boolean {
+        when (char) {
+            numberChar.ZERO.number, numberChar.ONE.number, numberChar.TWO.number,
+            numberChar.THREE.number, numberChar.FOUR.number, numberChar.FIVE.number,
+            numberChar.SIX.number, numberChar.SEVEN.number, numberChar.EIGHT.number,
+            numberChar.NINE.number -> return true
+            else -> return false
+        }
+    }
+
+    private fun caculateResults(): String {
+        val digitOperators = digitsOperator()
+        if (digitOperators.isEmpty()) return ""
+
+        val multiplyDivide = multiplyDivideCaculate(digitOperators)
+        if (multiplyDivide.isEmpty()) return ""
+
+        val result = addSubtractCaculate(multiplyDivide)
+        return result.toString()
+    }
+
+    private fun addSubtractCaculate(passedList: MutableList<Any>): Float {
+        var result = passedList[0].toString().toFloat()
```

Comment:
```suggestion
        var result = passedList.firstOrNull().toString().toFloat()
```
kotlin nó có thằng này nè em , tăng cường dùng null safety vì kotlin nó hỗ trợ nhiều